src/Components/WebView/WebView/src/IpcCommon.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Linq;
+using System.Text.Json;
+
+namespace Microsoft.AspNetCore.Components.WebView
+{
+    internal class IpcCommon
+    {
+        private const string _ipcMessagePrefix = "__bwv:";
+
+        public static string Serialize(OutgoingMessageType messageType, params object[] args)
+            => Serialize(messageType.ToString(), args);
+
+        public static string Serialize(IncomingMessageType messageType, params object[] args)
+            => Serialize(messageType.ToString(), args);
+
+        public static bool TryDeserializeIncoming(string message, out IncomingMessageType messageType, out ArraySegment<JsonElement> args)
+            => TryDeserialize(message, out messageType, out args);
+
+        public static bool TryDeserializeOutgoing(string message, out OutgoingMessageType messageType, out ArraySegment<JsonElement> args)
+            => TryDeserialize(message, out messageType, out args);
+
+        private static string Serialize(string messageType, object[] args)
+        {
+            // We could come up with something a little more low-level here if we
+            // wanted to avoid a couple of allocations
+            var messageTypeAndArgs = new object[] { messageType }.Concat(args);
+            return $"{_ipcMessagePrefix}{JsonSerializer.Serialize(messageTypeAndArgs, JsonSerializerOptionsProvider.Options)}";
+        }
+
+        private static bool TryDeserialize<T>(string message, out T messageType, out ArraySegment<JsonElement> args)
+        {
+            // We don't want to get confused by unrelated messages that the developer is sending
+            // over the same webview IPC channel, so ignore anything else
+            if (message != null && message.StartsWith(_ipcMessagePrefix, StringComparison.Ordinal))
+            {
+                var messageAfterPrefix = message.AsSpan(_ipcMessagePrefix.Length);
+                var parsed = (JsonElement[])JsonSerializer.Deserialize(messageAfterPrefix, typeof(JsonElement[]), JsonSerializerOptionsProvider.Options);
```

Comment:
Something to think about here is that we are double-serializing JSON data when we do JS Interop

Reply:
Yeah, I agree. I think we can make this interop mechanism more low-level and performant later if we want to. Hopefully it's all very much contained to this one file (and the equivalent on the `.ts` side) so it should be easy to reason about any changes.